### PR TITLE
st_collectionextract

### PR DESCRIPTION
**Pull request by ateucher**
I needed the equivalent of postgis' [`ST_CollectionExtract`](https://postgis.net/docs/ST_CollectionExtract.html) the other day, so I thought it might be worth offering it up. Definitely no hard feelings if you're not interested.  

I made it using only exported functions, so if there are more elegant/performant ways of doing this by going lower-level, let me know. I got the basics from [this stackoveflow question](https://stackoverflow.com/questions/45723775/how-to-select-certain-geometries-from-a-geometrycollection-after-st-intersect).

### CODE REVIEW

**Comment by ateucher** [comment]
I just found a case I didn't have a test for that doesn't work... working on a fix

File: R/collectionextract.R [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#' st_collectionextract(cc, "POLYGON")
+#'
+st_collectionextract = function(x, type = c("POLYGON", "POINT", "LINESTRING"), warn = FALSE) {
+	UseMethod("st_collectionextract")
+}
+
+#' @name st_collectionextract
+#'
+#' @export
+st_collectionextract.sfg = function(x, type = c("POLYGON", "POINT", "LINESTRING"), warn = FALSE) {
```
Comment (edzer) [inline R/collectionextract.R]: In line with function naming, please rename into `st_collection_extract` -- I think PostGIS users will understand, as we don't use camel case.

Comment (ateucher) [reply R/collectionextract.R]: Ok, will do

**Comment by ateucher** [comment]
I think that's better now

File: R/collection_extract.R [hunk 2]
```diff
@@ -0,0 +1,138 @@
+#' Given a (multi)geometry, return a (multi)geometry consisting only of elements of the specified type.
+#'
+#' Similar to st_collection_extract in postgis. If there are no sub-geometries of the specified type, an empty geometry is returned.
```
Comment (edzer) [inline R/collection_extract.R]: Similar to `ST_CollectionExtract` in PostGIS.

File: R/collection_extract.R [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+#'
+#' @param x an \code{sf(c/g)} object that has mixed geometry
```
Comment (edzer) [inline R/collection_extract.R]: an object of class \code{sf}, \code{sfc} or \code{sfg} that has...

With mixed geometry, do you also mean MULTIPOLYGON or only GEOMETRYCOLLECTION?

File: R/collection_extract.R [hunk 4]
```diff
@@ -0,0 +1,138 @@
+#' Given a (multi)geometry, return a (multi)geometry consisting only of elements of the specified type.
```
Comment (edzer) [inline R/collection_extract.R]: Also GEOMETRYCOLLECTION?

File: R/collection_extract.R [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+#'
+#' Similar to st_collection_extract in postgis. If there are no sub-geometries of the specified type, an empty geometry is returned.
+#'
+#' @param x an \code{sf(c/g)} object that has mixed geometry
+#' @param type One of "POLYGON", "POINT", "LINESTRING"
```
Comment (edzer) [inline R/collection_extract.R]: type character; one of

File: R/collection_extract.R [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+#' @param warn logical; if \code{TRUE}, warn if attributes are assigned to sub-geometries when casting (see \code{\link{st_cast}})
+#'
+#' @return An sf(c/g) object with (multi)geometries consisting only of elements of the specified type. For sfg objects, an sfg object is returned if there is only one geometry of the specified type, otherwise the geometries are combined into an sfc object of the relevant type. If any subgeometries in the input are MULTI, then all of the subgeometries in the output will be MULTI.
```
Comment (edzer) [inline R/collection_extract.R]: An object having the same class as \code{x}, with ...

sfg -> \code{sfg}

File: R/collection_extract.R [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	# Find the geometries of the specified type and extract into a list
+	matches = vapply(x, st_is, types, FUN.VALUE = logical(1))
+	x_types = x[which(matches)]
+	if (length(x_types) == 0L) {
+		## Not sure if this should be a warning and return an empty sfg (current)
+		## or error?
+		warning("x contains no geometries of specified type")
+		return(x_types)
```
Comment (edzer) [inline R/collection_extract.R]: better return `typed_empty(type)` here, so at least an `sfg` is returned?

File: R/collection_extract.R [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	} else if (length(x_types) == 1L) {
+		# Get the contents of the first (only) list element which is an sfg
+		return(x_types[[1]])
+	} else {
+		# turn list into an sfc, and cast it to single type. Will be multi
+		# if any are multi
+		return(st_cast(st_sfc(x_types), warn = warn))
```
Comment (edzer) [inline R/collection_extract.R]: Yeah, so we either return an sfg or an sfc from this method; I guess there's no better option.

**Comment by ateucher** [comment]
Great comments @edzer, thanks. I'll start working on them.

File: R/collection_extract.R (hunk 8 above)
Comment (ateucher) [reply R/collection_extract.R]: Not that I could find, I don't think running `st_multi()` on the list of `sfg` objects to force an `sfg` would be desirable.

File: R/collection_extract.R (hunk 3 above)
Comment (ateucher) [reply R/collection_extract.R]: As it's set up it works on only `GEOMETRYCOLLECTION` for `sfg` and `GEOMETRY` or `GEOMETRYCOLLECTION` for `sfc` and `sf`. It doesn't work to extract `POLYGON` from `MULTIPOLYGON` etc. I *think* this is consistent with `ST_CollectionExtract`. I'll try to make that clearer in the documentation.

File: R/collection_extract.R (hunk 7 above)
Comment (ateucher) [reply R/collection_extract.R]: Thanks. Didn't know about `typed_empty`

**Comment by edzer** [comment]
Thanks, great work!

**Comment by ateucher** [comment]
Fantastic, thanks!

**Pull request merged**